Summarize command

berndbohmeier opened the pull request:
This is the first iteration of the summarize command. It is by no mean done, but it has the most important first steps

UI:
- Sample Statistics
   answers which samples are done, which are failing. Gives an idea on how much the sequencing is already done
- QC Statistics
   gives an idea about how good the sequencing is working. Which experiments have good data, which have high coverage, which are contaminated, etc.
- Prevalence
   Two ways to show prevalence, grouped by columns in the metadata file

JasonAHendry committed:
Fix example script

The -e flag is not needed.

JasonAHendry committed:
Add functionality for cleaning sample_type column

JasonAHendry committed:
Add first-draft of summarise command

berndbohmeier committed:
Merge remote-tracking branch 'origin/main' into feature/summary

berndbohmeier committed:
Read in master metadata file

berndbohmeier committed:
Make summarize work with new format and delve

berndbohmeier committed:
Add sample summary statistic

berndbohmeier committed:
Add prevalence by region plot

berndbohmeier committed:
Small improvements to summarize

Get ready for better text
Ensure we mark duplicates and only take the best sample
Ensure we exclude samples not in master metadata

berndbohmeier committed:
Add samples amplicons barplot

berndbohmeier committed:
Add prevalence per region/year plot

berndbohmeier committed:
Fix filtering of false positives

berndbohmeier committed:
Update wsaf false positive threashold

berndbohmeier committed:
Move some files to utils experiment

berndbohmeier committed:
Add some more docs and move code in summarize

berndbohmeier committed:
Linter fixes and code structure

berndbohmeier committed:
Make sample_type mandatory

Make sample_type mandatory. We need it for summary analysis and it is
best if people just include it when creating the sample sheet. It should
not be much extra work.

We maybe want to make it optional if we can derive it from the sample
name, to be discussed.

berndbohmeier committed:
Do not seperate mutations by alt alleles

For now just don't group them by alt alleles. We might have a different
set of alt alleles when we call the same mutation in different
experiments, as we might have a triallelic side. As we in the end group
by amino acid change, this leads to problems. We need to better discuss
how to hanle csq calling correctly for multiple changes.

berndbohmeier committed:
Limit prevalence to samples in master metadata file

berndbohmeier committed:
Ensure we handle sample ids that are number better

berndbohmeier committed:
Store in inventory if samples are unknown

berndbohmeier committed:
Show legend in Sample statistic pie

This helps to only look at the sequenced samples

berndbohmeier committed:
Better text and labels for QC Summary

berndbohmeier committed:
New colorscales for QC summary

Green and red should be more intuitve what they mean

berndbohmeier committed:
Ensure prevalence is ordered by aa positions

berndbohmeier committed:
Always report contaminated when over abs. thresh.

It makes sense that we don't report contamination if we have low
coverage, as this could actually just be caused by low coverage. But if
we are over the abs threshold, we can be sure, it's contamination

berndbohmeier committed:
Allow to group by all columns in metadata file

Exclude columns with to many entries and which are numeric. Maybe we
want to filter more in the future, or have a way to provide a list

berndbohmeier committed:
Allow to start dashboard in debug mode via env

berndbohmeier committed:
Don't check for depth files in summary

They are big and we don't need them for the analysis

berndbohmeier committed:
Add gene deletion detection

berndbohmeier committed:
First version on map in summary

berndbohmeier committed:
Sort drug resistance markers by prevalence

berndbohmeier committed:
Use read mapping file instead of fastq file

Fastq files are optional so it is better to check a file that is
mandatory

berndbohmeier committed:
Do not require vcf folder at the moment

Burkina Faso does not have this folder. And as we don't use this folder
at the moment, let's not check it.

berndbohmeier committed:
Fix error message of wrong exp meta data file

berndbohmeier committed:
For gene deletion analysis, exclude contaminated samples

berndbohmeier committed:
Add site markers to map and settings file

berndbohmeier committed:
Add missing files

berndbohmeier committed:
Add show neg control coverage in summary

berndbohmeier committed:
Add option to create summary without master file

berndbohmeier committed:
Improve messages for summarize

berndbohmeier committed:
Fix bug in error bars of summary

berndbohmeier committed:
Remove problematic mutations again

berndbohmeier committed:
Add warnings to help with finding the right region names

berndbohmeier committed:
Add init file to summarize module

JasonAHendry commented on src/nomadic/util/metadata.py:
```diff
@@ -181,5 +264,37 @@ def _check_entries_unique(self):
         self.df["barcode"] = [correct_barcode_format(b) for b in self.df["barcode"]]
+
+
+class ExtendedMetadataTableParser(MetadataTableParser):
```
Previously I had `sample_type` optional for running `realtime`, and only required when wanting to do the `nomadic summarize`, using this subclass of `MetadataTableParser`.

Now that we are making `sample_type` mandatory, I think now this is unnecessary, and this code can be removed? 

However, using the subclass I also checked whether there was at least one positive and negative control in the experiment. This is important for some of the QC checks that are done in `summarize` (e.g. checking for contamination, and in the future it will be required for hrp2/3 deletion detection).

Is this done somewhere else now? Or how will `summarize` behave if the user has an experiment without a positive and/or negative control?

berndbohmeier replied:
It is checked that each experiment has at least one negative control. Currently having at least one positive control is not enforced.

JasonAHendry commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+        print("")
+        debug = bool(os.getenv("NOMADIC_DEBUG"))
+        if not debug:
+            webbrowser.open("http://127.0.0.1:8050")
```
I get a "Firefox can’t establish a connection to the server at 127.0.0.1:8050." with this implementation, which only corrects on refresh. 

Could we use the timer implementation in `nomadic realtime` e.g. lines 92-120 of `src/nomadic/realtime/dashboard/builders.py`?

If you want I can also make this commit, just let me know.

JasonAHendry commented on src/nomadic/summarize/dashboard/components.py:
```diff
@@ -0,0 +1,1014 @@
+                            [
+                                f"Experiments: {self.throughput_df.columns.shape[0] - 1}",
+                                html.Br(),
+                                f"Field samples (total): {self.throughput_df.loc['field', 'All']}",
```
Let's make these explicitly integers, so they don't display as e.g. `145.0` but `145`.

berndbohmeier committed:
Apply wording change for summary from code review

Co-authored-by: danieljbridges <51692943+danieljbridges@users.noreply.github.com>

berndbohmeier committed:
Clean up some things in summarize

danieljbridges commented on src/nomadic/summarize/dashboard/components.py:
```diff
@@ -0,0 +1,1014 @@
+
+
+SAMPLE_COLORS = {
+    "missing": "#636EFA",
```
Sounds good to me. I think that is more intuitive

JasonAHendry commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+    return variants_df
+
+
+def load_and_concat_variants(expt_dirs: list[str]) -> pd.DataFrame:
```
This needs to have differing behaviour for versions of Nomadic where the unfiltered VCF file is stored; it would be preferable to work from that when aggregating variants.

JasonAHendry commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+    master_metadata["sample_id"] = master_metadata["sample_id"].str.strip()
+
+    # In case sample IDs are numbers, we want to strip leading zeros (this was a problem in Zambia data)
+    inventory_metadata["sample_id"] = inventory_metadata["sample_id"].str.lstrip("0")
```
I don't think Nomadic should take on the responsibility of cleaning sample IDs. If the sample IDs have issues with them, we should leave that for users with knowledge of the study to fix.

JasonAHendry commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+        f"{output_dir}/summary.gene-deletions.prevalence.csv", index=False
+    )
+
+    for col in prevalence_by:
```
I think we should check that the `prevalence_by` columns are all in the metadata further up, and provide some graceful error handling if they are not.

JasonAHendry reviewed (changes requested):
OK -- I had a first look over the updated code.

I think we have a lot of great stuff here but still plenty of work to do to finalise. Some of the bigger things in my view are:

* thinking a bit more about what the 'core' summary command should provide; while allowing for some flexibility for bespoke analyses for specific panels (e.g. hrp2/3 deletions)
* reviewing how we handle mapping, and ideally providing a more generic solution that doesn't require the user providing GIS files themselves
* considering again how we handle behaviour across versions; e.g. what happens if we have some data processed with delve, and some with bcftools? Similarly, we should use the unfiltered VCF file if it is available
* allowing just the dashboard to be launched if the data has been processed already

See some comments throughout, can also talk more in person!

berndbohmeier committed:
Add pydantic to dependencies

berndbohmeier committed:
Rename missing to not_sequenced

berndbohmeier committed:
Use regex for matching of alt column names

berndbohmeier commented on src/nomadic/util/metadata.py:
```diff
@@ -103,6 +178,14 @@ class MetadataTableParser:
             "sample id",
             "sample ids",
         ],
+        "sample_type": [
```
done

berndbohmeier committed:
Also auto open browser after a delay for summary

This is because the dashboard needs some time to start

berndbohmeier committed:
Ensure dtype of throughput table is int

So we don't show 123.0 instead of 123

berndbohmeier committed:
Rename meta_data to metadata

berndbohmeier committed:
Remove summary command structure comment

berndbohmeier commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+    #
+    # --------------------------------------------------------------------------------
+
+    log.info("Calculate gene deletions...")
```
sure. Do you want me to delete it completly, because it is not helpful at all, or maybe keep it, just not enabled and with a comment that this is not very robust. I think it can still help to quickly find samples to resequence.

berndbohmeier committed:
Expose min coverage and max contamination values

berndbohmeier commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,805 @@
+
+    # Now let's evaluate coverage
+    coverage_df = get_region_coverage_dataframe(expt_dirs, inventory_metadata)
+    MIN_COV = 50
```
made them parameters of the CLI

JasonAHendry committed:
Expose false-positive filter and consolidate prev calc

Make the parameters used to define likely false-positive
mutations function arguments; consolidate two versions
of prevalence calculation functions.

JasonAHendry committed:
Simplify experiment coapletion checks

This commit does three things
1. Define a new class, called ExperimentOutputChecker,
that holds information about experiment outputs for
nomadic summarize.
2. Consolidate some functions that were duplicate (e.g.
find_metadata and get_metadata_csv
3. Reorganise the util/experiment.py module

JasonAHendry committed:
Remove ExperimentResultsChecker class

Replace with a function and dataclass

berndbohmeier committed:
A few typing fixes

berndbohmeier committed:
Remove fixing of leading zeros

We don't want to do this kind of fixing in nomadic, it should be done
before.

berndbohmeier committed:
Allow to only view with nomadic summarize command

We try it first like this, but imo it would be better to actually
separate viewing the summary into a different command or combine it with
dashboard. It is a bit awkward to combine an output command (produce the
files of summary) with an input command (showing files from a folder).

berndbohmeier committed:
Make summary general

Put all mvp specific setting into an object and make the dashboard work
for any panel with amplicons.

berndbohmeier committed:
Only plot data in summary if we have it

berndbohmeier committed:
Remove print statement

berndbohmeier committed:
Ensure to only include field samples in metadata

berndbohmeier committed:
Exit summary early if we have no field samples

berndbohmeier committed:
Ensure we print the experiment that has an metadata error

berndbohmeier committed:
More detailed output of what summary is loaded

berndbohmeier committed:
Ensure we set sample_id dtype when loading.

berndbohmeier committed:
Don't remove int columns for grouping

If we want to group by year for example, this is needed

berndbohmeier committed:
Ensure master metadata file for summary has no dup

berndbohmeier committed:
Fix gene deletions component

berndbohmeier committed:
Merge remote-tracking branch 'origin/main' into feature/summary

berndbohmeier committed:
Fix parser used for metadata

this was a mistake introduces in a merge

berndbohmeier committed:
Make sample_type not required again

After discussion we make sample_type column not required.
But we already normalize it in case it is present, so we split that
functionality.

berndbohmeier committed:
Merge remote-tracking branch 'origin/main' into feature/summary

berndbohmeier committed:
Update required version of bcftools

The reason is that --samples-list option for reheader was added here

berndbohmeier committed:
Use vcf files for amino acid change prevalence

We use the vcf files directly to have more precise information of
weather an amino acid change was not present or we had insufficient data
to tell.

berndbohmeier committed:
Add option to switch between bubble map and region map

berndbohmeier committed:
Sort aa changes by aa_pos not pos

berndbohmeier committed:
Ensure we read sample id as str

berndbohmeier committed:
Formatter fixes

berndbohmeier committed:
Update how maps are loaded

berndbohmeier committed:
Ensure we only load experiment dirs we need

We should only load variants and other data from an experiment that has
at least one included sample, otherwise we waste work

berndbohmeier committed:
Take out wrong asserts in summarize

They didn't really make sense as for additional groups we don't know if
they provide all data.

berndbohmeier committed:
Change default min coverage to 100

berndbohmeier committed:
Improve gene deletion code

berndbohmeier committed:
Set min number of samples to plot map to 10

berndbohmeier committed:
Formating

berndbohmeier committed:
Set filter false positive params for summarize

berndbohmeier committed:
Load shapeName for geojson from settings file

The feature part of geojson is not standardized. This allows to set a
different key for the name of the shape, e.g. the province name etc

berndbohmeier committed:
Merge remote-tracking branch 'origin/main' into feature/summary

berndbohmeier committed:
Make ExperimentDirectories side effect free

Ensure when creating a ExperimentDirectory object that it does not
directly create the dirs. This is wrong when opening the dashboard for
example

berndbohmeier committed:
Add auto find port function like in realtime to summary

berndbohmeier committed:
Refactored QC code of summarize

Put code into better maintainable structure.
- No pandas code in main (except still to_csv stuff)
- Every part of the analysis got it's own module
- Follow some best practices for pandas so that every function is pd.df
  -> pd.df

berndbohmeier commented on src/nomadic/summarize/analysis/qc.py:
```diff
@@ -0,0 +1,371 @@
+        )
+        .fillna({"n_replicates": 0, "n_passing": 0})
+        .astype({"n_replicates": int, "n_passing": int})
+    )
```
that is okay.

berndbohmeier commented on src/nomadic/summarize/dashboard/builders.py:
```diff
@@ -0,0 +1,595 @@
+from nomadic.util.summary import Settings, get_map_settings
+
+
+class SummaryDashboardBuilder(ABC):
```
I would leave it like this for now as we anyways plan to make a new UI

berndbohmeier commented on src/nomadic/summarize/main.py:
```diff
@@ -0,0 +1,472 @@
+    aa_changes_df, nt_changes_df = load_variants_from_vcfs(
+        expt_dirs,
+        caller=caller,
+        temp_dir=Path(output_dir) / "temp_vcf_processing",
```
can be supported later
